File: scripts/check-med-roundtrip.js
#!/usr/bin/env node
/* global __dirname */

const fs = require('fs');
const path = require('path');

const CASES_PATH = path.join(__dirname, 'med-roundtrip-cases.json');

function parseFormattedProfile(formattedText) {
  const metadata = [];
  const layers = [];
  const temperatures = [];
  const stabilityTests = [];
  const notes = [];

  const stabilityPattern =
    /\b(?:CT(?:E|M|H)?\d*(?:\s+(?:SP|SC|PC|RP|BRK))?\s+at\s+\d+(?:\.\d+)?\s*cm|ECT[PNX]?\d*(?:\s+(?:SP|SC|PC|RP|BRK))?\s*(?:at\s+\d+(?:\.\d+)?\s*cm)?|PST\s+\d+(?:\.\d+)?\/\d+(?:\.\d+)?(?:\s+(?:END|ARR|SF))?\s+at\s+\d+(?:\.\d+)?\s*cm|HS\s+(?:easy|moderate|hard)\s+at\s+\d+(?:\.\d+)?\s*cm|SS\s+(?:easy|moderate|hard)(?:\s+(?:SP|SC|PC|RP|BRK))?\s+at\s+\d+(?:\.\d+)?\s*cm|RB\d+\s+at\s+\d+(?:\.\d+)?\s*cm)\b/gi;

  const splitInlineStability = (line) => {
    const found = Array.from(line.matchAll(stabilityPattern)).map((m) => m[0].replace(/\s+/g, ' ').trim());
    const cleaned = line.replace(stabilityPattern, ' ').replace(/\s{2,}/g, ' ').trim();
    return { cleaned, tests: found };
  };

  formattedText
    .split('\n')
    .map((line) => line.trim())
    .filter(Boolean)
    .forEach((line) => {
      const metadataMatch = line.match(/^([^:]+):\s*(.+)$/);
      if (metadataMatch) {
        metadata.push({ label: metadataMatch[1], value: metadataMatch[2] });
        return;
      }
      if (/^-?\d+(?:\.\d+)?\s+(?:surface|\d+(?:\.\d+)?cm)$/i.test(line)) {
        temperatures.push(line);
        return;
      }
      if (/^(CT|ECT|PST|HS|SS|DT|RB)/i.test(line)) {
        const split = splitInlineStability(line);
        (split.tests.length > 0 ? split.tests : [line]).forEach((test) => stabilityTests.push(test));
        return;
      }
      if (/^\d+(?:\.\d+)?-\d+(?:\.\d+)?\s+/i.test(line)) {
        const split = splitInlineStability(line);
        layers.push(/^\d+(?:\.\d+)?-\d+(?:\.\d+)?\s+/i.test(split.cleaned) ? split.cleaned : line);
        split.tests.forEach((test) => stabilityTests.push(test));
        return;
      }
      notes.push(line);
    });

  return { metadata, layers, temperatures, stabilityTests, notes };
}

function parseLegacyLayerLine(line) {
  const match = line.trim().match(/^(\d+(?:\.\d+)?)-(\d+(?:\.\d+)?)\s+(.+)$/);
  if (!match) return null;
  const top = match[1];
  const bottom = match[2];
  const [main, comment] = match[3].split('|').map((s) => s.trim());
  const isRed = /\bred\b/i.test(main);
  const tokens = main
    .replace(/\bred\b/gi, '')
    .replace(/(\d+(?:\.\d+)?)\s*(mm|cm)\b/gi, '$1$2')
    .replace(/\s*\/\s*/g, '/')
    .replace(/\s+/g, ' ')
    .trim()
    .split(' ')
    .filter(Boolean);

  const grain = tokens.find((t) => /^(PP|DF|RG|FC|FCxr|SH|DH|MF|IF|IFrc|MFcr)(\/(PP|DF|RG|FC|FCxr|SH|DH|MF|IF|IFrc|MFcr))?$/i.test(t)) || '';
  const hardness = tokens.find((t) => /^(F|F\+|4F|4F\+|1F|1F\+|P|P\+|K|K\+|I)(-(F|F\+|4F|4F\+|1F|1F\+|P|P\+|K|K\+|I))?$/i.test(t)) || '';
  const size = tokens.find((t) => /^\d+(?:\.\d+)?(mm|cm)(\/\d+(?:\.\d+)?(mm|cm))?$/i.test(t)) || '';
  return { top, bottom, grain, hardness, size, comment: comment || '', concern: isRed ? 'yes' : 'no' };
}

function parseStabilityLine(line) {
  const normalized = line.replace(/,/g, ' ').replace(/\s+/g, ' ').trim();
  const upper = normalized.toUpperCase();
  const type = upper.startsWith('ECT')
    ? 'ECT'
    : upper.startsWith('CT')
      ? 'CT'
      : upper.startsWith('PST')
        ? 'PST'
        : upper.startsWith('SS')
          ? 'SS'
          : upper.startsWith('HS')
            ? 'HS'
            : upper.startsWith('RB')
              ? 'RB'
              : '';
  if (!type) return null;

  const test = {
    type,
    depth: (normalized.match(/\bat\s+(\d+(?:\.\d+)?)\s*cm\b/i) || [])[1] || '',
    taps: (normalized.match(/\b(\d{1,2})\s*taps?\b/i) || [])[1] || '',
    character: ((normalized.match(/\b(SC|SP|PC|RP|BRK)\b/i) || [])[1] || '').toUpperCase(),
    result: '',
    pstCut: '',
    pstColumn: '',
  };

  if (type === 'CT') {
    const ctCode = (normalized.match(/\bCT([EMH])\b/i) || [])[1];
    if (ctCode === 'E' || ctCode === 'e') test.result = 'easy';
    if (ctCode === 'M' || ctCode === 'm') test.result = 'moderate';
    if (ctCode === 'H' || ctCode === 'h') test.result = 'hard';
    if (!test.taps) {
      const compact = normalized.match(/\bCT(?:E|M|H)?\s*(\d{1,2})\b/i);
      if (compact && compact[1]) test.taps = compact[1];
    }
  }
  if (type === 'ECT') {
    const ect = normalized.match(/\bECT([NPX])(\d{1,2})?\b/i);
    if (ect && ect[1]) test.result = `ECT${ect[1].toUpperCase()}`;
    if (ect && ect[2] && !test.taps) test.taps = ect[2];
  }
  if (type === 'PST') {
    const pst = normalized.match(/\b(END|ARR|SF)\b/i);
    if (pst && pst[1]) test.result = pst[1].charAt(0).toUpperCase() + pst[1].slice(1).toLowerCase();
    const col = normalized.match(/\b(\d+(?:\.\d+)?)\s*\/\s*(\d+(?:\.\d+)?)\b/);
    if (col) {
      test.pstCut = col[1];
      test.pstColumn = col[2];
    }
  }
  if (type === 'RB') {
    const rb = normalized.match(/\b(RB[1-7])\b/i);
    if (rb && rb[1]) test.result = rb[1].toUpperCase();
  }

  return test;
}

function serializeLayer(layer) {
  const main = [`${layer.top}-${layer.bottom}`, layer.grain, layer.hardness, layer.size, layer.concern === 'yes' ? 'red' : '']
    .filter(Boolean)
    .join(' ')
    .trim();
  return layer.comment ? `${main} | ${layer.comment}` : main;
}

function serializeStability(test) {
  if (test.type === 'CT') {
    const resultCode = test.result === 'easy' ? 'CTE' : test.result === 'moderate' ? 'CTM' : test.result === 'hard' ? 'CTH' : 'CT';
    const tapsPart = test.taps ? `${test.taps} taps` : '';
    const charPart = test.character || '';
    const depthPart = test.depth ? `at ${test.depth} cm` : '';
    return [resultCode, tapsPart, charPart, depthPart].filter(Boolean).join(' ').replace(/\s+/g, ' ').trim();
  }
  if (test.type === 'ECT') {
    const resultPart = test.result || 'ECT';
    const tapsPart = test.taps || '';
    const charPart = test.character || '';
    const depthPart = test.depth ? `at ${test.depth} cm` : '';
    return [resultPart, tapsPart, charPart, depthPart].filter(Boolean).join(' ').replace(/\s+/g, ' ').trim();
  }
  if (test.type === 'PST') {
    const ratio = test.pstCut && test.pstColumn ? `${test.pstCut}/${test.pstColumn}` : '';
    const resultPart = test.result ? test.result.toUpperCase() : '';
    const depthPart = test.depth ? `at ${test.depth} cm` : '';
    return ['PST', ratio, resultPart, depthPart].filter(Boolean).join(' ').replace(/\s+/g, ' ').trim();
  }
  if (test.type === 'RB') {
    const resultPart = test.result || 'RB';
    const depthPart = test.depth ? `at ${test.depth} cm` : '';
    return [resultPart, depthPart].filter(Boolean).join(' ').trim();
  }
  const depthPart = test.depth ? `at ${test.depth} cm` : '';
  return [test.type, test.result, test.taps ? `${test.taps} taps` : '', test.character, depthPart]
    .filter(Boolean)
    .join(' ')
    .replace(/\s+/g, ' ')
    .trim();
}

function normalizeLine(line) {
  return line.replace(/\s+/g, ' ').trim().toUpperCase();
}

function runCase(testCase) {
  const parsed = parseFormattedProfile(testCase.formattedText);
  const hydratedLayers = parsed.layers.map(parseLegacyLayerLine).filter(Boolean);
  const hydratedTests = parsed.stabilityTests.map(parseStabilityLine).filter(Boolean);
  const reserializedLayers = hydratedLayers.map(serializeLayer);
  const reserializedTests = hydratedTests.map(serializeStability);

  const layerLoss = reserializedLayers.length !== parsed.layers.length;
  const testLoss = reserializedTests.length !== parsed.stabilityTests.length;
  const layerDiff = reserializedLayers.some((line, i) => normalizeLine(line) !== normalizeLine(parsed.layers[i] || ''));

  return {
    name: testCase.name,
    pass: !(layerLoss || testLoss || layerDiff),
    details: {
      inputLayers: parsed.layers.length,
      outputLayers: reserializedLayers.length,
      inputTests: parsed.stabilityTests.length,
      outputTests: reserializedTests.length,
      layerDiff,
    },
  };
}

function main() {
  const raw = fs.readFileSync(CASES_PATH, 'utf8');
  const cases = JSON.parse(raw);
  const results = cases.map(runCase);
  const failed = results.filter((r) => !r.pass);

  console.log(`MED roundtrip regression: ${cases.length} case(s)`);
  for (const result of results) {
    if (result.pass) {
      console.log(`[PASS] ${result.name}`);
    } else {
      console.log(`[FAIL] ${result.name} ${JSON.stringify(result.details)}`);
    }
  }

  if (failed.length > 0) {
    process.exit(1);
  }
  console.log('All MED roundtrip checks passed.');
}

main();

File: scripts/check-med-roundtrip.py
#!/usr/bin/env python3
import json
import re
from pathlib import Path

CASES_PATH = Path(__file__).with_name("med-roundtrip-cases.json")


def parse_formatted_profile(formatted_text: str):
    metadata, layers, temperatures, stability_tests, notes = [], [], [], [], []
    stability_pattern = re.compile(
        r"\b(?:CT(?:E|M|H)?\d*(?:\s+(?:SP|SC|PC|RP|BRK))?\s+at\s+\d+(?:\.\d+)?\s*cm|"
        r"ECT[PNX]?\d*(?:\s+(?:SP|SC|PC|RP|BRK))?\s*(?:at\s+\d+(?:\.\d+)?\s*cm)?|"
        r"PST\s+\d+(?:\.\d+)?/\d+(?:\.\d+)?(?:\s+(?:END|ARR|SF))?\s+at\s+\d+(?:\.\d+)?\s*cm|"
        r"HS\s+(?:easy|moderate|hard)\s+at\s+\d+(?:\.\d+)?\s*cm|"
        r"SS\s+(?:easy|moderate|hard)(?:\s+(?:SP|SC|PC|RP|BRK))?\s+at\s+\d+(?:\.\d+)?\s*cm|"
        r"RB\d+\s+at\s+\d+(?:\.\d+)?\s*cm)\b",
        re.I,
    )

    for line in [x.strip() for x in formatted_text.splitlines() if x.strip()]:
        md = re.match(r"^([^:]+):\s*(.+)$", line)
        if md:
            metadata.append({"label": md.group(1), "value": md.group(2)})
            continue
        if re.match(r"^-?\d+(?:\.\d+)?\s+(?:surface|\d+(?:\.\d+)?cm)$", line, re.I):
            temperatures.append(line)
            continue
        if re.match(r"^(CT|ECT|PST|HS|SS|DT|RB)", line, re.I):
            found = [re.sub(r"\s+", " ", m.group(0)).strip() for m in stability_pattern.finditer(line)]
            (found if found else [line]).__iter__()
            for test in (found if found else [line]):
                stability_tests.append(test)
            continue
        if re.match(r"^\d+(?:\.\d+)?-\d+(?:\.\d+)?\s+", line, re.I):
            found = [re.sub(r"\s+", " ", m.group(0)).strip() for m in stability_pattern.finditer(line)]
            cleaned = re.sub(stability_pattern, " ", line)
            cleaned = re.sub(r"\s{2,}", " ", cleaned).strip()
            layers.append(cleaned if re.match(r"^\d+(?:\.\d+)?-\d+(?:\.\d+)?\s+", cleaned, re.I) else line)
            for test in found:
                stability_tests.append(test)
            continue
        notes.append(line)

    return {
        "metadata": metadata,
        "layers": layers,
        "temperatures": temperatures,
        "stability_tests": stability_tests,
        "notes": notes,
    }


def parse_legacy_layer_line(line: str):
    m = re.match(r"^(\d+(?:\.\d+)?)-(\d+(?:\.\d+)?)\s+(.+)$", line.strip())
    if not m:
        return None
    top, bottom, tail = m.group(1), m.group(2), m.group(3)
    parts = [p.strip() for p in tail.split("|", 1)]
    main = parts[0]
    comment = parts[1] if len(parts) > 1 else ""
    is_red = bool(re.search(r"\bred\b", main, re.I))
    tokens = (
        re.sub(r"\s+", " ", re.sub(r"\s*/\s*", "/", re.sub(r"(\d+(?:\.\d+)?)\s*(mm|cm)\b", r"\1\2", re.sub(r"\bred\b", "", main, flags=re.I))))
        .strip()
        .split(" ")
    )
    grain = next((t for t in tokens if re.match(r"^(PP|DF|RG|FC|FCxr|SH|DH|MF|IF|IFrc|MFcr)(/(PP|DF|RG|FC|FCxr|SH|DH|MF|IF|IFrc|MFcr))?$", t, re.I)), "")
    hardness = next((t for t in tokens if re.match(r"^(F|F\+|4F|4F\+|1F|1F\+|P|P\+|K|K\+|I)(-(F|F\+|4F|4F\+|1F|1F\+|P|P\+|K|K\+|I))?$", t, re.I)), "")
    size = next((t for t in tokens if re.match(r"^\d+(?:\.\d+)?(mm|cm)(/\d+(?:\.\d+)?(mm|cm))?$", t, re.I)), "")
    return {"top": top, "bottom": bottom, "grain": grain, "hardness": hardness, "size": size, "comment": comment, "concern": "yes" if is_red else "no"}


def parse_stability_line(line: str):
    normalized = re.sub(r"\s+", " ", line.replace(",", " ")).strip()
    up = normalized.upper()
    if up.startswith("ECT"):
        t = "ECT"
    elif up.startswith("CT"):
        t = "CT"
    elif up.startswith("PST"):
        t = "PST"
    elif up.startswith("SS"):
        t = "SS"
    elif up.startswith("HS"):
        t = "HS"
    elif up.startswith("RB"):
        t = "RB"
    else:
        return None

    depth = (re.search(r"\bat\s+(\d+(?:\.\d+)?)\s*cm\b", normalized, re.I) or [None, ""])[1]
    taps = (re.search(r"\b(\d{1,2})\s*taps?\b", normalized, re.I) or [None, ""])[1]
    character = ((re.search(r"\b(SC|SP|PC|RP|BRK)\b", normalized, re.I) or [None, ""])[1]).upper()
    result, pst_cut, pst_column = "", "", ""

    if t == "CT":
        code = (re.search(r"\bCT([EMH])\b", normalized, re.I) or [None, ""])[1].upper()
        result = {"E": "easy", "M": "moderate", "H": "hard"}.get(code, "")
        if not taps:
            compact = re.search(r"\bCT(?:E|M|H)?\s*(\d{1,2})\b", normalized, re.I)
            if compact:
                taps = compact.group(1)
    elif t == "ECT":
        ect = re.search(r"\bECT([NPX])(\d{1,2})?\b", normalized, re.I)
        if ect:
            result = f"ECT{ect.group(1).upper()}"
            if ect.group(2) and not taps:
                taps = ect.group(2)
    elif t == "PST":
        pst = re.search(r"\b(END|ARR|SF)\b", normalized, re.I)
        if pst:
            result = pst.group(1).capitalize()
        col = re.search(r"\b(\d+(?:\.\d+)?)\s*/\s*(\d+(?:\.\d+)?)\b", normalized)
        if col:
            pst_cut, pst_column = col.group(1), col.group(2)
    elif t == "RB":
        rb = re.search(r"\b(RB[1-7])\b", normalized, re.I)
        if rb:
            result = rb.group(1).upper()

    return {"type": t, "depth": depth, "taps": taps, "character": character, "result": result, "pst_cut": pst_cut, "pst_column": pst_column}


def serialize_layer(layer):
    main = " ".join([x for x in [f"{layer['top']}-{layer['bottom']}", layer["grain"], layer["hardness"], layer["size"], "red" if layer["concern"] == "yes" else ""] if x]).strip()
    return f"{main} | {layer['comment']}".strip() if layer["comment"] else main


def normalize_line(s: str):
    return re.sub(r"\s+", " ", s).strip().upper()


def run_case(case):
    parsed = parse_formatted_profile(case["formattedText"])
    hydrated_layers = [x for x in (parse_legacy_layer_line(line) for line in parsed["layers"]) if x]
    hydrated_tests = [x for x in (parse_stability_line(line) for line in parsed["stability_tests"]) if x]
    out_layers = [serialize_layer(x) for x in hydrated_layers]
    layer_loss = len(out_layers) != len(parsed["layers"])
    test_loss = len(hydrated_tests) != len(parsed["stability_tests"])
    layer_diff = any(normalize_line(out_layers[i]) != normalize_line(parsed["layers"][i]) for i in range(min(len(out_layers), len(parsed["layers"]))))
    return {
        "name": case["name"],
        "pass": not (layer_loss or test_loss or layer_diff),
        "details": {
            "input_layers": len(parsed["layers"]),
            "output_layers": len(out_layers),
            "input_tests": len(parsed["stability_tests"]),
            "output_tests": len(hydrated_tests),
            "layer_diff": layer_diff,
        },
    }


def main():
    cases = json.loads(CASES_PATH.read_text())
    results = [run_case(c) for c in cases]
    failed = [r for r in results if not r["pass"]]

    print(f"MED roundtrip regression: {len(cases)} case(s)")
    for result in results:
        if result["pass"]:
            print(f"[PASS] {result['name']}")
        else:
            print(f"[FAIL] {result['name']} {result['details']}")

    if failed:
        raise SystemExit(1)
    print("All MED roundtrip checks passed.")


if __name__ == "__main__":
    main()
